### PR TITLE
net/can/: add statistics for recv, sent and drop

### PR DESCRIPTION
**Pull request by occam25**
## Summary

Add support for network statistics for CAN. It includes counters for received, sent and dropped frames.
This change was necessary for our application to monitor the dropped can frames reported in the field.
The implementation has mimicked the existing ones for tcp and udp statistics, adding a can_stats_s struct to store the recv, sent and drop counters with the already existing net_stats_t type.

## Impact

With this change, users can access the CAN statistics through procfs, the same way they access to all the other network statistics. The CAN recv, sent, and drop counters have been added to the /proc/net/stat output following the same design already existing for other network devices. This change doesn't change any driver behavior beyond just adding the new statistics. It does not introduce any new dependencies beyond the needed header files needed to access the new can_stats_s struct and g_netstats variable. These includes are nuttx/net/netconfig.h and nuttx/net/can.h.
The CAN statistics implemented are only enabled if CONFIG_NET_STATISTICS and CONFIG_NET_CAN configurations are set.

## Testing

Tested using a Ubuntu 22.04 and a stm32h7 custom board with CONFIG_NET_STATISTICS and CONFIG_NET_CAN configurations enabled. The test application opens a CAN interface using socketCAN and echoes all the frames read. Every two seconds we print out the CAN statistics and see the recv and sent counters increase accordingly. To test the drop counter, we stop reading the CAN interface from the application so the frames start accumulating in the network buffer. Eventually, that buffer gets full and we start seeing the drop counter increasing as expected.
This traces are printed after reading the counters from /proc/net/stat
CAN stats:
   Received: 00000022
   Sent:     00000022
   Dropped:  00000010

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No, this PR description does not fully meet the NuttX requirements.  Here's why and what's missing:

* **Summary:** While it states the *what*, it lacks the *why* and *how*.
    * **Missing:**  Why is this change necessary? Is it for a specific application, to improve monitoring, or to address a deficiency?  How exactly does the change implement the statistics tracking?  Are new syscalls added?  Are existing data structures modified?
    * **Needs:**  More detail on the implementation approach.

* **Impact:**  Almost entirely missing.  This section needs to be filled out thoughtfully.
    * **Missing:** All impact assessments.
    * **Needs:**  Consider each point.  For example, impact on the user is likely YES as they now have access to new information. Describe how users can access the statistics (new ioctl calls, procfs entries, etc.). Does it impact build size?  Does it change any driver behavior beyond just adding statistics? Does it introduce any new dependencies?

* **Testing:** Insufficient.
    * **Missing:**  "Testing logs before change" and "Testing logs after change" are empty.  Also, "stm32h7 custom board" is too vague.
    * **Needs:** Provide actual log output demonstrating the statistics working. Be specific about the board and configuration used (e.g., "stm32h743-nucleo with the `nsh` config"). Show the commands used to generate the statistics and how the statistics are retrieved. Show logs *before* the change to demonstrate that the stats were not available and logs *after* the change showing the stats being collected and retrieved.

In short, the PR description needs to be significantly more detailed to meet the NuttX requirements.  It needs to provide the missing *why* and *how* in the summary, complete the impact assessment, and include actual test logs demonstrating the change in functionality.

File: net/procfs/net_statistics.c [hunk 1]
```diff
@@ -40,7 +40,7 @@
 
 #if defined(CONFIG_NET_IPv4) || defined(CONFIG_NET_IPv6) || \
     defined(CONFIG_NET_TCP) || defined(CONFIG_NET_UDP) || \
-    defined(CONFIG_NET_ICMP) || defined(CONFIG_NET_ICMPv6)
+    defined(CONFIG_NET_ICMP) || defined(CONFIG_NET_ICMPv6) || defined(CONFIG_NET_CAN)
```
Comment (acassis) [inline net/procfs/net_statistics.c:43]: Please move defined(CONFIG_NET_CAN) to the next line to keep two per line, or change previous line to keep three per line

**Comment by acassis** [comment]
Nice work Javier! Really well done PR and very detailed PR summary! Kudos!!!

File: net/procfs/net_statistics.c (hunk 1 above)
Comment (occam25) [reply net/procfs/net_statistics.c]: Done

File: include/nuttx/net/can.h [hunk 2]
```diff
@@ -43,6 +44,7 @@
 
 #include <nuttx/config.h>
 #include <nuttx/can.h>
+#include <nuttx/net/netconfig.h>
```
Comment (xiaoxiang781216) [inline include/nuttx/net/can.h:47]: why add this

File: net/can/can_callback.c [hunk 3]
```diff
@@ -236,6 +234,7 @@ uint16_t can_datahandler(FAR struct net_driver_s *dev,
   else
     {
       nerr("ERROR: Failed to queue the I/O buffer chain: %d\n", ret);
+      ret = 0;
```
Comment (xiaoxiang781216) [inline net/can/can_callback.c:237]: why zero ret

File: include/nuttx/net/can.h (hunk 2 above)
Comment (occam25) [reply include/nuttx/net/can.h:47]: there is where net_stats_t type (needed in can_stats_s new structure) is defined

File: net/can/can_callback.c (hunk 3 above)
Comment (occam25) [reply net/can/can_callback.c:237]: This was actually a bug. This function returns an uint16_t as the number of bytes actually buffered, but the call to iob_tryadd_queue (line 218)  returns an int and, if it fails, it returns a negative value. In that case, can_datahandler return value is cast to an unsigned value as if it would had buffered many bytes. 
Setting the returning value to zero when iob_tryadd_queue returns a negative value reflects the actual meaning that no bytes have been buffered

Comment (xiaoxiang781216) [reply net/can/can_callback.c:237]: it's better to move this change to a new patch and give a clear commit message

File: net/procfs/net_statistics.c (hunk 1 above)
Comment (xiaoxiang781216) [reply net/procfs/net_statistics.c]: let's fix the problem in the same patch. @occam25

Comment (occam25) [reply net/procfs/net_statistics.c]: I guess you mean leaving the `ret = 0` fix discussed in the other comment to be part of this PR. I agree, otherwise the CAN statistics implemented in this patch would not work properly

File: net/can/can_callback.c (hunk 3 above)
Comment (occam25) [reply net/can/can_callback.c:237]: I think this should be part of this PR, otherwise the CAN statistics implemented in this patch would not work properly. I can create a specific commit for this patch if you prefer

Comment (occam25) [reply net/can/can_callback.c:237]: done @xiaoxiang781216

**Review by acassis — APPROVED** [review]

**Review by acassis — APPROVED** [review]
